### PR TITLE
New Sync to LMS flow using FormAssembly entered Salesforce data - Core Logic

### PR DESCRIPTION
**Pull request by sadleb**
Implements core functionality of the new Sync To LMS flow. There are TODOs littered throughout with corresponding Asana tasks for a future iteration.

# High level context

## The old Sync To LMS flow was:
1. Fellow or LC creates an account on the Join server
2. fills out an application
3. we store that information locally and push most of it to Salesforce
4. regional staff review, accept, confirm and map them to a cohort on Salesforce
5. regional staff hit a Sync To LMS button on Salesforce which calls into the Join server with the Campaign ID to sync
6. Join server pulls info from the Salesforce Campaign and combines it with information stored locally to hit the Canvas API and provision the users.

## The new flow for getting fully confirmed and enrolled Fellows and LCs into the Portal:
1. Fellow or LC fills out a FormAssembly form in Salesforce which populates the new Salesforce "BluePrint" data model
2. Regional staff funnel their application through to the point of creating a Participant object in a Program which represents someone who is fully ready to have Portal access and participate in the program
3. A nightly cron job provisions Participants in the Portal with the correct access
4. An on-demand Sync To LMS button is also available on Salesforce which kicks off the sync cronjob for that Program (aka course)
5. A launch logistics email goes out with a link to create a Braven account and get Portal access (the email must match)
6.  When they create the account, as long as Sync To LMS has been run they get Portal access

# How to test / Try this out
1. See the `README` for more info about setting up Salesforce. But here is the test Program that I created in our staging environment 
image
2. Clicking on the Fellows tab you can see that I added an Enrolled Participant 
image
3. Did the same for Leadership Coaches.
4. Assuming you have the Salesforce ENV vars set in your shell, you can go here to run the sync [http://platformweb:3020/salesforce/sync_to_lms] by putting in the course ID that is set as the `Target Course In LMS` field on your test Salesforce Program
image
5. You can also run it using a rake task. E.g. `bundle exec rake salesforce:sync_to_lms[71]`
6. Also, if you don't want to deal with your Canvas dev environment, you can change the `CANVAS_URL` and `CANVAS_TOKEN` ENV vars to the staging values.

### CODE REVIEW

File: config/environments/development.rb [hunk 1]
```diff
@@ -17,14 +17,15 @@
   if Rails.root.join('tmp', 'caching-dev.txt').exist?
     config.action_controller.perform_caching = true
 
-    config.cache_store = :memory_store
+    config.cache_store = :file_store, "tmp/cache"
```
[the remark this answers is not in the document]
Comment (sadleb) [reply config/environments/development.rb]: Heroku deploys were failing b/c of this. Need to install the 'dalli' gem and a memcachestore add-on. Punting on enabling the Rails.cache until we actually need it.

**Comment by rshipp** [comment]
Tested this with course 10 (the test course) and it gave a 500 - not sure if this matters, I'm assuming it's because 10 doesn't have any enrolled users in SF.

If it helps, this is the traceback

```
2020-04-17T16:03:10.092748+00:00 app[web.1]: I, [2020-04-17T16:03:10.092633 #4]  INFO -- : [7cd61020-04a9-4a5c-89f8-a6f436621c58] Started POST "/salesforce/sync_to_lms" for 70.241.96.132 at 2020-04-17 16:03:10 +0000
2020-04-17T16:03:10.094095+00:00 app[web.1]: I, [2020-04-17T16:03:10.093999 #4]  INFO -- : [7cd61020-04a9-4a5c-89f8-a6f436621c58] Processing by SalesforceController#sync_to_lms as HTML
2020-04-17T16:03:10.094175+00:00 app[web.1]: I, [2020-04-17T16:03:10.094109 #4]  INFO -- : [7cd61020-04a9-4a5c-89f8-a6f436621c58]   Parameters: {"utf8"=>"✓", "authenticity_token"=>"", "course_id"=>"10", "commit"=>"Save "}
2020-04-17T16:03:10.096962+00:00 app[web.1]: D, [2020-04-17T16:03:10.096884 #4] DEBUG -- : [7cd61020-04a9-4a5c-89f8-a6f436621c58]   [1m[36mUser Load (0.7ms)[0m  [1m[34mSELECT "users".* FROM "users" WHERE "users"."id" = $1 ORDER BY "users"."id" ASC LIMIT $2[0m  [["id", 1], ["LIMIT", 1]]
2020-04-17T16:03:10.098073+00:00 app[web.1]: I, [2020-04-17T16:03:10.097982 #4]  INFO -- : [7cd61020-04a9-4a5c-89f8-a6f436621c58] Completed 500 Internal Server Error in 4ms (ActiveRecord: 0.7ms | Allocations: 918)
2020-04-17T16:03:10.098209+00:00 app[web.1]: D, [2020-04-17T16:03:10.098142 #4] DEBUG -- : [7cd61020-04a9-4a5c-89f8-a6f436621c58] undefined method `+' for nil:NilClass excluded from capture: DSN not set
2020-04-17T16:03:10.098743+00:00 app[web.1]: F, [2020-04-17T16:03:10.098670 #4] FATAL -- : [7cd61020-04a9-4a5c-89f8-a6f436621c58]   
2020-04-17T16:03:10.098744+00:00 app[web.1]: [7cd61020-04a9-4a5c-89f8-a6f436621c58] NoMethodError (undefined method `+' for nil:NilClass):
2020-04-17T16:03:10.098745+00:00 app[web.1]: [7cd61020-04a9-4a5c-89f8-a6f436621c58]   
2020-04-17T16:03:10.098747+00:00 app[web.1]: [7cd61020-04a9-4a5c-89f8-a6f436621c58] lib/salesforce_api.rb:16:in `initialize'
2020-04-17T16:03:10.098747+00:00 app[web.1]: [7cd61020-04a9-4a5c-89f8-a6f436621c58] lib/sync_to_lms.rb:12:in `new'
2020-04-17T16:03:10.098748+00:00 app[web.1]: [7cd61020-04a9-4a5c-89f8-a6f436621c58] lib/sync_to_lms.rb:12:in `initialize'
2020-04-17T16:03:10.098748+00:00 app[web.1]: [7cd61020-04a9-4a5c-89f8-a6f436621c58] app/controllers/salesforce_controller.rb:15:in `new'
2020-04-17T16:03:10.098749+00:00 app[web.1]: [7cd61020-04a9-4a5c-89f8-a6f436621c58] app/controllers/salesforce_controller.rb:15:in `sync_to_lms'
```

**Comment by sadleb** [comment]
> Tested this with course 10 (the test course) and it gave a 500 - not sure if this matters, I'm assuming it's because 10 doesn't have any enrolled users in SF.
> 
> Edit: also fails with the same 500 when I use course 71 (testing in the review app).
> 
> If it helps, this is the traceback
> 
> ```
> 2020-04-17T16:03:10.092748+00:00 app[web.1]: I, [2020-04-17T16:03:10.092633 #4]  INFO -- : [7cd61020-04a9-4a5c-89f8-a6f436621c58] Started POST "/salesforce/sync_to_lms" for 70.241.96.132 at 2020-04-17 16:03:10 +0000
> 2020-04-17T16:03:10.094095+00:00 app[web.1]: I, [2020-04-17T16:03:10.093999 #4]  INFO -- : [7cd61020-04a9-4a5c-89f8-a6f436621c58] Processing by SalesforceController#sync_to_lms as HTML
> 2020-04-17T16:03:10.094175+00:00 app[web.1]: I, [2020-04-17T16:03:10.094109 #4]  INFO -- : [7cd61020-04a9-4a5c-89f8-a6f436621c58]   Parameters: {"utf8"=>"✓", "authenticity_token"=>"", "course_id"=>"10", "commit"=>"Save "}
> 2020-04-17T16:03:10.096962+00:00 app[web.1]: D, [2020-04-17T16:03:10.096884 #4] DEBUG -- : [7cd61020-04a9-4a5c-89f8-a6f436621c58]   �[1m�[36mUser Load (0.7ms)�[0m  �[1m�[34mSELECT "users".* FROM "users" WHERE "users"."id" = $1 ORDER BY "users"."id" ASC LIMIT $2�[0m  [["id", 1], ["LIMIT", 1]]
> 2020-04-17T16:03:10.098073+00:00 app[web.1]: I, [2020-04-17T16:03:10.097982 #4]  INFO -- : [7cd61020-04a9-4a5c-89f8-a6f436621c58] Completed 500 Internal Server Error in 4ms (ActiveRecord: 0.7ms | Allocations: 918)
> 2020-04-17T16:03:10.098209+00:00 app[web.1]: D, [2020-04-17T16:03:10.098142 #4] DEBUG -- : [7cd61020-04a9-4a5c-89f8-a6f436621c58] undefined method `+' for nil:NilClass excluded from capture: DSN not set
> 2020-04-17T16:03:10.098743+00:00 app[web.1]: F, [2020-04-17T16:03:10.098670 #4] FATAL -- : [7cd61020-04a9-4a5c-89f8-a6f436621c58]   
> 2020-04-17T16:03:10.098744+00:00 app[web.1]: [7cd61020-04a9-4a5c-89f8-a6f436621c58] NoMethodError (undefined method `+' for nil:NilClass):
> 2020-04-17T16:03:10.098745+00:00 app[web.1]: [7cd61020-04a9-4a5c-89f8-a6f436621c58]   
> 2020-04-17T16:03:10.098747+00:00 app[web.1]: [7cd61020-04a9-4a5c-89f8-a6f436621c58] lib/salesforce_api.rb:16:in `initialize'
> 2020-04-17T16:03:10.098747+00:00 app[web.1]: [7cd61020-04a9-4a5c-89f8-a6f436621c58] lib/sync_to_lms.rb:12:in `new'
> 2020-04-17T16:03:10.098748+00:00 app[web.1]: [7cd61020-04a9-4a5c-89f8-a6f436621c58] lib/sync_to_lms.rb:12:in `initialize'
> 2020-04-17T16:03:10.098748+00:00 app[web.1]: [7cd61020-04a9-4a5c-89f8-a6f436621c58] app/controllers/salesforce_controller.rb:15:in `new'
> 2020-04-17T16:03:10.098749+00:00 app[web.1]: [7cd61020-04a9-4a5c-89f8-a6f436621c58] app/controllers/salesforce_controller.rb:15:in `sync_to_lms'
> ```

Just for other, we discussed this over slack. It's not a problem. The SALESFORCE ENV vars weren't setup.

File: app/controllers/salesforce_controller.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
+require 'sync_to_lms'
+
+class SalesforceController < ApplicationController
+
+  def sync_to_lms
+    # TODO: this is a complete hack just to get a prototype going. When we implement this for real,
```
Comment (hcli-work) [inline app/controllers/salesforce_controller.rb:6]: turned these into new user stories

**Pull request merged**